Fix Jenkins.getInstance() to make life easier for plugin authors

stephenc opened the pull request:
@jenkinsci/code-reviewers @reviewbybees

![sad panda2](https://cloud.githubusercontent.com/assets/209336/13616921/e4ebdf6e-e573-11e5-8f06-5563c14a9a56.png)

The original change to annotate `Jenkins.getInstance()` with `@CheckForNull` was a mistake:

Nobody did an actual trace of execution flow for plugins. 
- Plugins should be safe to assume that `Jenkins.getInstance() != null` on any code path that they should encounter... e.g. plugins doing wrong things like registering a shutdown hook _and_ not deregistering it when the Jenkins is shut down are broken and need fixing
- 95% of the usage in core is perfectly safe to make the same assumption as plugins.
- There is one remaining edge case in plugins, namely when code is sent over remoting and contains a reference to the `Jenkins` class... those plugins are equally broken and need fixing too...

The net result is that all the plugins since fa496c7bf228ecb243acf0135106f18f27132c6f have been forced to make a pointless change from `Jenkins.getInstance()` to `Jenkins.getActiveInstance()`.

The correct change should have been to introduce a `Jenkins.getInstanceOrNull()` for the 5% of uses where it is valid to have a null instance on the execution path.

What we now have without this change is lots of plugin authors confused as to whether they should:
- use `getInstance()` and try and figure out how to respond to it being null
- use `getActiveInstance()` and try and figure out what to do if it throws an exception because the backing instance is null

So we are making life harder for plugin developers just to fix a single failing unit test without changing the test and... well I cannot see what else.

Once this is merged we should then merge #2092 to tighten up the calling contract.

Merging this change should be a high priority as there is a cost to repeatedly updating the PR as other PRs get merged potentially introducing new calls to `getInstance()` that need analysis and subsequent review

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented:
You will also need to clarify the deprecation reason of <code>getActiveInstance()</code>, because the current suggestion is a bit misleading

stephenc committed:
Jenkins.getInstance() will be @Nonnull in Jenkins 2.0+

- Code that is running from a plugin and on the master's JVM is guaranteed to never get null from this method (any cases where you do get null are bugs in core)
- Code that is running from a plugin and on a remote JVM should never be allowed to load the Jenkins class in their classloader, so should never use Jenkins.getInstance()... we are annotating the method with @Nullable so that such code can have some evolution time
- Code that is running in core and on one of two special paths should use the Jenkins.getInstanceOrNull() method so that the UI can be presented to users before the singleton has been instantiated / after the singleton has been destroyed
- The remaining 95% of uses in core (and 100% of uses in plugins) can safely assume that the instance is never null

stephenc commented:
@jenkinsci/code-reviewers @reviewbybees ping

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -706,10 +709,27 @@ protected File getRootDirFor(String name) {
+     * In other cases you may have code that might end up running on a remote JVM and not on the Jenkins master,
+     * for those cases you really should rewrite your code so that when the {@link Callable} is sent over the remoting
+     * channel it uses a {@code writeReplace} method or similar to ensure that the {@link Jenkins} class is not being
+     * loaded into the remote class loader
```
I believe a fix of [JENKINS-26677](https://issues.jenkins-ci.org/browse/JENKINS-26677) would help diagnose this mistake.

jglick replied:
(IOW, I agree that calling `getInstanceOrNull` is the wrong fix for such cases: they should be fixed to not refer to `Jenkins` to begin with. There are some crazy idioms like [this](https://github.com/jenkinsci/subversion-plugin/blob/57460ce3cfb0885fce878c47eafec45e83e2fc00/src/main/java/hudson/scm/SubversionWorkspaceSelector.java#L105-L108) that caused headaches for SECURITY-144 anyway.)

stephenc replied:
Yes either a fix for that or throwing `IllegalStateException` from `getInstance()` ;-)

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -706,10 +709,27 @@ protected File getRootDirFor(String name) {
+
+    /**
+     * Gets the {@link Jenkins} singleton. In certain rare cases you may have code that is intended to run before
+     * Jenkins starts or while Jenkins is being shut-down. For those rare cases use {@link #getInstanceOrNull()}.
```
Not so much that the code is _intended_ to run while Jenkins is being shut down, as that it runs asynchronously, and `cleanUp()` is not exhaustive, so on occasion it could run after shutdown, until you find a way to ensure that it does not.

stephenc replied:
Well we should just fix any cases where `cleanUp()` is not exhaustive. 

If you are calling e.g. `Jenkins.getInstance()` or `ExtensionList.lookup` after `cleanUp` has cleared `theInstance` then essentially you will result in a partial re-bootstrap of Jenkins itself and a failure to cleanUp correctly

jglick commented:
A modest proposal:

```diff
diff --git a/core/src/main/java/jenkins/model/Jenkins.java b/core/src/main/java/jenkins/model/Jenkins.java
index 7fcf37c..43e6552 100644
--- a/core/src/main/java/jenkins/model/Jenkins.java
+++ b/core/src/main/java/jenkins/model/Jenkins.java
@@ -758,8 +758,6 @@ public class Jenkins extends AbstractCIBase implements DirectlyModifiableTopLeve
             this.root = root;
             this.servletContext = context;
             computeVersion(context);
-            if(theInstance!=null)
-                throw new IllegalStateException("second instance");
             theInstance = this;
 
             if (!new File(root,"jobs").exists()) {
@@ -2883,8 +2881,6 @@ public class Jenkins extends AbstractCIBase implements DirectlyModifiableTopLeve
             }
 
         LogFactory.releaseAll();
-
-        theInstance = null;
     }
 
     public Object getDynamic(String token) {
```

(The first hunk to allow `RestartableJenkinsRule` to work.) Would solve a lot of hard-to-reproduce errors. Does this actually break anything? (Besides, perhaps, CJP HA, which you or I could fix.)

jglick commented:
:bee: in that I see no harm done by this PR, and it would ease migration of baseline version for workaday plugins which have no code run in exotic circumstances (i.e., outside of web methods or builders). I am less optimistic that the hard-line change can be done in 2.0 without causing a number of regressions (generally under timing-dependent circumstances, therefore hard to reproduce and fix).

stephenc commented:
> I am less optimistic that the hard-line change can be done in 2.0

My thinking is that for 2.0's first release we just turn on the throws in unit tests (or perhaps in plugins that have been compiled against Jenkins 2.0+ if I can find an easy way to determine that... when running in a remote classloader... but if we get [JENKINS-26677](https://issues.jenkins-ci.org/browse/JENKINS-26677) then that issue goes away anyway)

Then after a while we can turn it on for all code paths... but given some people's comments on the javadoc changes I'd rather have the javadoc contract say 2.0+ now

jglick commented:
> for 2.0's first release we just turn on the throws in unit tests

Perhaps it would highlight cases that were previously _detected_ during functional tests and worked around with a null check. Would probably not catch the cases that are ultimately the result of asynchronous calls and were detected in some other manner.

jtnord commented:
:bee:

stephenc commented:
@reviewbybees done

stephenc commented:
I'll resolve the merge conflicts shortly

reviewbybees commented:
This pull request has completed our internal processes and we now respectfully request the maintainers of this repository to consider our proposal contained within this pull request for merging.

stephenc committed:
Merge remote-tracking branch 'jenkins/master' into alt-pr-2090

stephenc commented:
@oleg-nenashev @KostyaSha @andresrc @batmat as you all commented on bb7c8fcedbcc9b51c5b1bb5b32810af5ac6b1ffb are we good to merge?

jglick commented:
Some of the test failures seem to stem from the infamous undiagnosed

    java.lang.NoClassDefFoundError: Could not initialize class com.cloudbees.plugins.credentials.SystemCredentialsProvider$ProviderImpl

But [this](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/4200/testReport/hudson.model/DownloadService2Test/loadJSONHTML/), [this](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/4200/testReport/hudson.model/DownloadServiceTest/testLoadJSON/), and [this](https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/4200/testReport/hudson.model/UpdateCenterTest/data/) seem to be real regressions:

```
java.lang.NullPointerException: null
	at hudson.ProxyConfiguration._get(ProxyConfiguration.java:278)
	at hudson.ProxyConfiguration.get(ProxyConfiguration.java:270)
	at hudson.ProxyConfiguration.open(ProxyConfiguration.java:220)
	at …
```

stephenc commented:
@jglick I think those failures are actually from how the test harness sets up jenkins outside of WebAppMain... https://github.com/jenkinsci/jenkins/pull/2104 should fix them

stephenc commented:
OK, I'm going to go ahead, resolve the merge conflict and merge

stephenc committed:
Merge remote-tracking branch 'jenkins/master' into alt-pr-2090
